Disallow users to bypass BC breaking changes

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72314
* #72313

Differential Revision: [D33997712](https://our.internmc.facebook.com/intern/diff/D33997712)

tugsbayasgalan committed:
Disallow users to bypass BC breaking changes

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997712).

mruberry commented on docs/guidance_for_upgraders.md:
```diff
@@ -0,0 +1,69 @@
+# Guidance for Operator Developer
+
+After you change either the operator signature in a BC breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `
```
This is not always true. Prototype and beta modules don't have BC-breaking guarantees and upgraders are only required if the operator is actually used by a torchscript or mobile model.

gmagogsfm replied:
Ah, good to know. What is an example of such a module? Is there a canonical source of truth where we can query whether a module is beta and not dispatchable to TorchScript?

mruberry commented on test/forward_backward_compatibility/check_forward_backward_compatibility.py:
```diff
@@ -227,8 +102,8 @@ def check_bc(existing_schemas):
     else:
         print(
             "The PR is introducing backward incompatible changes to the "
```
Again we can't block developers from making BC-breaking changes in coordination with operators. This test also checks FC, too, which upgraders don't deal with.

gmagogsfm replied:
What does "in coordination with operators" mean? The breaking change is on operator, so it is always in coordination with itself. 

Indeed, we can't deal with FC breaking changes now, it still has to follow 2 week wait policy, we we need at least an allowlist for FC.

tugsbayasgalan committed:
Update on "Disallow users to bypass BC breaking changes"


Differential Revision: [D33997712](https://our.internmc.facebook.com/intern/diff/D33997712)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33997712).

mruberry commented:
Per offline discussion we decided to think about our approach to testing BC/FC more generally before removing the allow list.
